core/commands/dag/dag.go

```diff
@@ -347,4 +357,69 @@ The output of blocks happens in strict DAG-traversal, first-seen, order.
 
 		return <-errCh
 	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+
+			var showProgress bool
+			val, specified := res.Request().Options[progressOptionName]
+			if !specified {
+				// default based on TTY availability
+				errStat, _ := os.Stderr.Stat()
+				if 0 != (errStat.Mode() & os.ModeCharDevice) {
+					showProgress = true
+				}
+			} else if val.(bool) {
+				showProgress = true
+			}
+
+			// simple passthrough
+			if !showProgress {
+				return cmds.Copy(re, res)
+			}
+
+			// We will be supplied a reader, simply print out how many bytes were passed as a crude progress
+
+			var exportSize int64
+			buf := make([]byte, 4*1024*1024) // this does not mean "read 4MB at a time", rather make a buffer larger than any reasonable pipe size
+			defer func() {
+				if exportSize > 0 {
+					os.Stderr.WriteString("\n")
+				}
+			}()
+
+			for {
+				v, err := res.Next()
+				if err != nil {
+					return re.CloseWithError(err)
+				}
+
+				if r, ok := v.(io.Reader); ok {
+					// we got a reader passed as a response
+					// proxy it through with an increasing counter
+					for {
+						len, readErr := r.Read(buf)
+						if len > 0 {
+							if err := re.Emit(bytes.NewBuffer(buf[:len])); err != nil {
```

Comment:
Side note:

* In this case, you'd actually want https://golang.org/pkg/bytes/#NewReader. `bytes.NewBuffer` works, but is overkill.
* Alternatively, you could use (and keep reusing) a https://golang.org/pkg/io/#LimitedReader, resetting `N` each time.

However, can we just use `progressBarForReader` from the `get`/`cat` commands? IIRC, we're allowed to set the size to 0.

Reply:
> can we just use progressBarForReader 

Making it so

Reply:
@Stebalien, @willscott  ok, I force pushed the PR. It is indeed much neater this way. I did not copy any of the "clear STDERR" crazy from the `commands` package, I am not sure why we are jumping through so many hoops over there

In action:
( while retrieval runs )
```
admin:~/devel/go-ipfs$ cmd/ipfs/ipfs dag export QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm | sha256sum
 13s  12.89 MiB / ? [-------=-----------------------------------------------------------------------] 972.67 KiB/s 
```

when done:
```
cmd/ipfs/ipfs dag export QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm | sha256sum
 1s  107.08 MiB / ? [-------------------------------------------=---------------------------------] 99.08 MiB/s 1s 
5a49df60c4842e7fe71849410025a8a5a98e2d0e6e71f90b4913635f7df7250d  -

```

Reply:
>  I did not copy any of the "clear STDERR" crazy from the commands package, I am not sure why we are jumping through so many hoops over there

Run an `add` command and look at the output. Without the clearing, we'd see something like:

```
added QmFoobar----------------------------] 99.08MiB/s 1s
added QmOther----------------------------] 99.08MiB/s 1s
...
```

Reply:
Ah... that's only on recursion though. Each piece that got submitted on the CLI *should* get its own bar / sizing that doesn't disappear on me after.

Adding it as a note when I get a chance to come back and clean up some of that, post-dagger.

Reply:
> Ah... that's only on recursion though. Each piece that got submitted on the CLI should get its own bar / sizing that doesn't disappear on me after.

Note: `ipfs add` uses a single bar for _everything_, calculating the total size up-front.

Reply:
I see... :sadpanda: